crypto/pkcs11key/key.go

```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
 	if !gotModulus || !gotExponent {
 		return noKey, errors.New("public key missing either modulus or exponent")
 	}
-	return rsa.PublicKey{
+
+	rsa := rsa.PublicKey{
 		N: n,
 		E: e,
-	}, nil
+	}
+	return rsa, nil
+}
+
+// Get the public key matching an Elliptic Curve private key
+func (ps *Key) getECPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) (interface{}, error) {
+	var noKey interface{}
+
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_EC_PARAMS, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, nil),
+	}
+
+	attr, err := module.GetAttributeValue(session, privateKeyHandle, template)
+	if err != nil {
+		return noKey, err
+	}
+
+	oid := []byte {}
+	id := []byte {}
+	gotOid, gotID := false, false
+	for _, a := range attr {
+		if a.Type == pkcs11.CKA_EC_PARAMS {
+			oid = a.Value
+			gotOid = true
+		} else if a.Type == pkcs11.CKA_ID {
+			id = a.Value
+			gotID = true
+		}
+	}
+	if !gotOid || !gotID {
+		return noKey, errors.New("public key missing either curve parameters or id")
+	}
+
+	poid := new(asn1.ObjectIdentifier)
+	asn1.Unmarshal(oid, poid)
+	curve := namedCurveFromOID(*poid)
+
+	templateSearch := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, id),
+	}
+
+	if err := module.FindObjectsInit(session, templateSearch); err != nil {
+		return noKey, err
+	}
+	objs, _, err := module.FindObjects(session, 2)
```

Comment:
This is something I meant to look into / fix in the RSA code: the second parameter of `FindObjects` is `max`, the max number of objects to be returned. I think it should be one, since we only care about one object. Would you mind fixing that here and in the RSA branch?